bibblio.py

```diff
@@ -0,0 +1,377 @@
+    ]
+
+    BIBBLIO_TEXT_LIMIT = 200000
+    TEXT_MEDIA_TYPES = [
+        Representation.TEXT_PLAIN,
+        Representation.TEXT_HTML_MEDIA_TYPE,
+    ]
+
+    def __init__(self, _db, custom_list_identifier,
+                 api=None, catalogue_identifier=None, fiction=False):
+        output_source = DataSource.lookup(_db, self.DATA_SOURCE_NAME)
+        super(BibblioCoverageProvider, self).__init__(
+            self.SERVICE_NAME, [], output_source,
+            batch_size=self.DEFAULT_BATCH_SIZE, operation=self.OPERATION
+        )
+
+        self.custom_list = CustomList.find(
+            self._db, DataSource.LIBRARY_STAFF, custom_list_identifier
+        )
+        self.fiction = fiction
+        self.api = api or BibblioAPI.from_config(self._db)
+        self.catalogue_id = catalogue_identifier
+
+    def items_that_need_coverage(self, identifiers=None):
+        data_sources = [DataSource.lookup(self._db, ds)
+                        for ds in self.INSTANT_CLASSICS_SOURCES]
+
+        # Get any uncovered editions from approved sources.
+        edition_subquery = Edition.missing_coverage_from(
+            self._db, data_sources, self.output_source,
+            operation=self.operation).subquery()
+
+        # Get any identifiers with uncovered editions in the targeted
+        # CustomList and an associated Work to be recommended.
+        qu = self._db.query(Identifier)\
+            .join(Identifier.primarily_identifies).join(Edition.work)\
+            .join(Edition.custom_list_entries).join(CustomListEntry.customlist)\
+            .join(edition_subquery, Edition.id == edition_subquery.c.id)\
+            .filter(CustomList.id==self.custom_list.id)
+
+        if not self.fiction:
+            # Only get nonfiction. This is the default setting.
+            qu = qu.filter(Work.fiction==False)
+
+        if identifiers:
+            qu = qu.filter(Id.id.in_([i.id for i in identifiers]))
+
+        return qu
+
+    def process_item(self, identifier, force=False):
+        content_item = self.content_item_from_identifier(identifier)
+
+        try:
+            result = self.api.create_content_item(content_item)
+        except Exception as e:
+            return CoverageFailure(
+                identifier, str(e), data_source=self.output_source,
+                transient=True
+            )
+
+        content_item_id = result.get('contentItemId')
+        bibblio_identifier, _is_new = Identifier.for_foreign_id(
+            self._db, Identifier.BIBBLIO_CONTENT_ITEM_ID, content_item_id
+        )
+
+        identifier.equivalent_to(
+            self.output_source, bibblio_identifier, 1
+        )
+        return identifier
+
+    def add_coverage_record_for(self, item):
+        """Adds an appropriate CoverageRecord for this particular
+        identifier and any equivalent identifiers.
+
+        Because Bibblio is text-based recommendation source, it doesn't
+        make sense to represent the same work multiple times.
```

Comment:
This tells me it might make more sense to make this a WorkCoverageProvider instead.